pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,406 @@
+	baseDiff := makeBaseDiff(old, new)
+	if baseDiff != Undecided {
+		propDiff := baseDiff.ToPropertyDiff()
+		if propDiff == nil {
+			return nil, nil
+		}
+		if differ.isForceNew(path) || mapHasReplacements(diff) {
+			propDiff = promoteToReplace(propDiff)
+		}
+		return map[detailedDiffKey]*pulumirpc.PropertyDiff{path.Key(): propDiff}, nil
+	}
+	return nil, errors.New("diff is not simplified")
+}
+
+func (differ detailedDiffer) makeTopPropDiff(
+	path propertyPath, old, new resource.PropertyValue,
+) map[detailedDiffKey]*pulumirpc.PropertyDiff {
+	baseDiff := makeBaseDiff(old, new)
+	isForceNew := differ.isForceNew(path)
+	var propDiff *pulumirpc.PropertyDiff
+	if baseDiff != Undecided {
+		propDiff = baseDiff.ToPropertyDiff()
+	} else if !old.DeepEquals(new) {
+		propDiff = &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE}
+	}
+
+	if isForceNew {
+		propDiff = promoteToReplace(propDiff)
+	}
+
+	if propDiff != nil {
+		return map[detailedDiffKey]*pulumirpc.PropertyDiff{path.Key(): propDiff}
+	}
+	return nil
+}
+
+func (differ detailedDiffer) makePropDiff(
+	path propertyPath, old, new resource.PropertyValue,
+) map[detailedDiffKey]*pulumirpc.PropertyDiff {
+	if path.IsReservedKey() {
+		return nil
+	}
+	propType := differ.getEffectiveType(differ.propertyPathToSchemaPath(path))
+
+	switch propType {
+	case shim.TypeList:
+		return differ.makeListDiff(path, old, new)
+	case shim.TypeSet:
+		// TODO[pulumi/pulumi-terraform-bridge#2200]: Implement set diffing
+		return differ.makeListDiff(path, old, new)
+	case shim.TypeMap:
+		return differ.makeObjectDiff(path, old, new)
+	default:
+		return differ.makeTopPropDiff(path, old, new)
+	}
+}
+
+func (differ detailedDiffer) makeListDiff(
+	path propertyPath, old, new resource.PropertyValue,
+) map[detailedDiffKey]*pulumirpc.PropertyDiff {
+	diff := make(map[detailedDiffKey]*pulumirpc.PropertyDiff)
+	oldList := []resource.PropertyValue{}
+	newList := []resource.PropertyValue{}
+	if isPresent(old) && old.IsArray() {
+		oldList = old.ArrayValue()
+	}
+	if isPresent(new) && new.IsArray() {
+		newList = new.ArrayValue()
+	}
+
+	// naive diffing of lists
+	// TODO[pulumi/pulumi-terraform-bridge#2295]: implement a more sophisticated diffing algorithm
+	// investigate how this interacts with force new - is identity preserved or just order
+	longerLen := max(len(oldList), len(newList))
+	for i := 0; i < longerLen; i++ {
+		elemKey := path.Index(i)
+		oldOk := i < len(oldList)
+		oldVal := resource.NewNullProperty()
+		if oldOk {
+			oldVal = oldList[i]
+		}
+		newOk := i < len(newList)
+		newVal := resource.NewNullProperty()
+		if newOk {
+			newVal = newList[i]
+		}
+
+		d := differ.makePropDiff(elemKey, oldVal, newVal)
+		for subKey, subDiff := range d {
+			diff[subKey] = subDiff
+		}
+	}
+
+	simplerDiff, err := differ.simplifyDiff(diff, path, old, new)
+	if err == nil {
+		return simplerDiff
+	}
+
+	return diff
+}
+
+func (differ detailedDiffer) makeObjectDiff(
```

Comment:
Quick comment here that this seems to handle both maps and objects (different things in the schema) OK.